### PR TITLE
Allow userOwner to be null in search-backend-reindex command

### PR DESCRIPTION
**Pull request by bramstroker**
## Changes in this pull request  

Fixes another case where `userOwner` could be null causing broken functionality.

During `pimcore:search-backend-reindex` following exception occurs when the indexed element has `userOwner` set to null, causing the whole reindex command to crash.

```
[2023-09-07T13:37:00.134630+02:00] console.CRITICAL: Error thrown while running command "pimcore:search-backend-reindex". Message: "Cannot assign null to property Pimcore\Bundle\SimpleBackendSearchBundle\Model\Search\Backend\Data::$userOwner of type int" {"exception":"[object] (TypeError(code: 0): Cannot assign null to property Pimcore\\Bundle\\SimpleBackendSearchBundle\\Model\\Search\\Backend\\Data::$userOwner of type int at /home/zitmaxx/domains/pim-zitmaxx.emico.nl/application/releases/601/vendor/pimcore/pimcore/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:336)","command":"pimcore:search-backend-reindex","message":"Cannot assign null to property Pimcore\\Bundle\\SimpleBackendSearchBundle\\Model\\Search\\Backend\\Data::$userOwner of type int"} []
[2023-09-07T13:37:00.149121+02:00] php.CRITICAL: Uncaught Error: Cannot assign null to property Pimcore\Bundle\SimpleBackendSearchBundle\Model\Search\Backend\Data::$userOwner of type int {"exception":"[object] (TypeError(code: 0): Cannot assign null to property Pimcore\\Bundle\\SimpleBackendSearchBundle\\Model\\Search\\Backend\\Data::$userOwner of type int at /vendor/pimcore/pimcore/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:336)"}
```

Also this field is nullable in the DB `search_backend_data`, so the model also reflects this correctly now.

## Remarks

Btw I branched from 11.x because this is only an issue introduced in pimcore 11, and the separate SimpleBackendSearchBundle didn't exist before.

### CODE REVIEW

**Comment by blankse** [comment]
You should rebase it to branch `11.0` and change the target branch to `11.0` because its a bug fix.

**Comment by bramstroker** [comment]
Should be ok now.

File: bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php [hunk 1]
```diff
@@ -97,7 +97,7 @@ class Data extends AbstractModel
      *
      * @var int
```
Comment (blankse) [inline bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php]:
```suggestion
     * @var int|null
```

File: bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php [hunk 2]
```diff
@@ -95,9 +95,9 @@ class Data extends AbstractModel
     /**
      * User-ID of the owner
      *
-     * @var int
+     * @var int|null
      */
-    protected int $userOwner;
+    protected ?int $userOwner = null;
```
Comment (kingjia90) [inline bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: should the setter also be `nullable`?
image
to reproduce do this
image
and search for `Jaguar` in the backend search (🔍 > DataObject) or quick search

https://github.com/pimcore/pimcore/blob/d982fa8d96dc775e49a04017cfa657882cedf148/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data/Dao.php#L83

Comment (bramstroker) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: I did have a look into other setters for nullable fields, for example `setUserModification`, `setData` and `setCreationDate` etc. before issueing this PR. 
Did not change the `setUserOwner` signature to keep it consistent with the others.
The code you referenced is not calling any setters.

With the code in this MR the search indexer works fine for 1.5 million objects and assets.
Except that we get an unserialisation error on an asset crashing the command. Will have a look into that next week.

It makes sense to me to allow a field with a value to be set to `null`, thus changing the signature of the setters. This is an architectonic question I suppose. I suggest to do that in other PR, and maybe also look into the other data objects if they have similar issues.

Comment (bramstroker) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: Hmm having a closer look at your comment, I see you were able to reproduce issue on quicksearch with the setter called with `null`.
I will change the signature. Do I also have to change the others to be consistent?

Comment (bramstroker) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: I have changed setters of `modificationDate`, `userModification` and `userOwner`

Comment (blankse) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: @bramstroker @kingjia90 Should we also change the setter of `creationDate`?

Comment (kingjia90) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: @bramstroker Thank you for the changes!
@blankse probably yes, i've took a quick look, overall all the `setCreationDate()` in models are not accepting `null` but the getter usually do. The `search_backend_data` DAO due to its purpose and usage, it is "relying/tied" on these nullable getters.

Comment (kingjia90) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: Alternatively could it be fixed by null coalescing all the values in these setters
https://github.com/pimcore/pimcore/blob/91a8f819595f06d00758033d2e066357ca33635c/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data/Listing/Dao.php#L50-L61

Comment (bramstroker) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: `null` coalescing will not work for `userModification`, `userOwner` and `modificationData` as these are expected to be nullable, also for the search backend table, but the source elements as well. They can have a `userOwner` set to `null` when the user is removed for example.

I don't know why the creationData is nullable in the first place. Imo this should be not null on database level in the first place.
I would suggest to leave the setter of creationDate as is, because when it is set to `null` this is not expected and will cause other issues down the road. Now it will result in an error while saving, which is better.

**Comment by kingjia90** [comment]
LGTM but there's something at even bigger picture that doesn't look right at the moment.

~Feels like we are loosening something that is meant to be stricter, and that on paper, `0` and `null` are just meant to be the same as `system`, right? There are no any purposeful distinction for that.~
0 is system, `null` could be deleted user

If we consider this PR valid, probably even here (and as result the ElementInterface), the setters should be also nullable (BC-Break)?
https://github.com/pimcore/pimcore/blob/4dc3b3e4280fb5cc92293224bff19c56ccd21444/models/Element/AbstractElement.php#L170-L180

the following is the code from 10.6, the phpdoc hint and the typecast are conflicting
https://github.com/pimcore/pimcore/blob/dd81ef4c666b18c254333867a60f6ed455025076/models/Element/AbstractElement.php#L241-L251

**Comment by kingjia90** [comment]
Feeling that the solution would be a migration to change all user related column from `null` to `0` in `search_backend_data` and null coalescing when "copying" the element's data into it https://github.com/pimcore/pimcore/blob/53449686033cdb9dd10f5c64c9636b8c725e2a35/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php#L326-L347

Update: also haven't figured it out yet but this `$after` is set to 0 somewhere
MicrosoftTeams-image (2)

File: bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php (hunk 2 above)
Comment (kingjia90) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: Looks like the suggested nullable setter example here was wrong, i changed it manually on the DB to simulate a `null` value persisted, but wasn't aware of the *magic* on  `save()` as in the screenshot below, the value would be always set and when not, would be 0.
https://user-images.githubusercontent.com/6014195/267583592-9dc926e5-c697-4a2f-b2da-96581e23c839.png

The fact that it crashes on non-nullable property is something that we are facing recently with the unserialized elements, so could be just that case.

So in short, only the changes in the first commit were correct. @bramstroker  Could you please revert the setter change?

Comment (bramstroker) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: @kingjia90 I have removed last two commits by force push.
Should be ok now.

I don't have time to deep dive into this.

But one thing to consider is `0` is the system user id. Which is correct when the system user actually created or modified the object.
`null` however is another case, because that's the scenario when a user is removed from the system, therefore we cannot know the user anymore. Setting it to 0 is imo just wrong, because than it would be listed as system user in the frontend, which might cause confusion.
Not a really big deal I think, but it's something to consider.

I suggest to merge this PR as it at least solves the search reindex for now, which is currently broken for our installation and probably also for others using Pimcore 11.
Then you pimcore core devs can decide how to solve this case throughout pimcore codebase and take next steps.

Comment (kingjia90) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: Thank you, the point of deleted users is something that i didn't take in account at all, i would take a deeper look into it as well.

Comment (kingjia90) [reply bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:100]: merging this Pr to the moment and would open a new issue to discuss about this null by deletion

**Comment by bramstroker** [comment]
@kingjia90 Unfortunately our QA tester found another case where we have similar issue.

The error occurs when loading a grid in the admin.

```
Timestamp: Thu Sep 14 2023 08:32:18 GMT+0200 (Central European Summer Time)
Status: 500 | 
URL: /admin/bundle/search/search/find?xaction=read&_dc=1694673138069
Method: POST
Message: Pimcore\Bundle\SimpleBackendSearchBundle\Model\Search\Backend\Data::setUserModification(): Argument #1 ($userModification) must be of type int, null given, called in /var/domains/x/application/releases/605/vendor/pimcore/pimcore/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data/Listing/Dao.php on line 58
Trace: 
in /var/domains/x/application/releases/605/vendor/pimcore/pimcore/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data.php:251
#0 /var/domains/x/application/releases/605/vendor/pimcore/pimcore/bundles/SimpleBackendSearchBundle/src/Model/Search/Backend/Data/Listing/Dao.php(58): Pimcore\Bundle\SimpleBackendSearchBundle\Model\Search\Backend\Data->setUserModification(NULL)
#1 [internal function]: Pimcore\Bundle\SimpleBackendSearchBundle\Model\Search\Backend\Data\Listing\Dao->load()
#2 /var/domains/x/application/releases/605/vendor/pimcore/pimcore/lib/Model/AbstractModel.php(222): call_user_func_array(Array, Array)
#3 /var/domains/x/application/releases/605/vendor/pimcore/pimcore/bundles/SimpleBackendSearchBundle/src/Controller/SearchController.php(326): Pimcore\Model\AbstractModel->__call('load', Array)
#4 /var/domains/x/application/releases/605/vendor/symfony/http-kernel/HttpKernel.php(182): Pimcore\Bundle\SimpleBackendSearchBundle\Controller\SearchController->findAction(Object(Symfony\Component\HttpFoundation\Request), Object(Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher), Object(Pimcore\Bundle\AdminBundle\Helper\GridHelperService))
#5 /var/domains/x/application/releases/605/vendor/symfony/http-kernel/HttpKernel.php(76): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#6 /var/domains/x/application/releases/605/vendor/symfony/http-kernel/Kernel.php(197): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#7 /var/domains/x/application/releases/605/public/index.php(36): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#8 {main}
```

This is upon fetching the data from `search_backend_table` and hydrating the `Search\Backend\Data` models.

Imo the right solution is that all getters/setters in this model matches their DB counterparts nullable schema.

Screenshot 2023-09-14 at 08 54 40

For us changing the signature from:

`public function setUserOwner(int $userOwner): static`

to

`public function setUserOwner(?int $userOwner): static`

Solves the issue.

I have made a matrix inventarising the issue and probable solution.

```
| field            | DB nullable | getter nullable | setter nullable | preferred solution     |
|------------------|-------------|-----------------|-----------------|------------------------|
| id               | N           | Y               | Y               |                        |
| fullpath         | Y           | N               | N               | DB migration           |
| maintype         | N           | None            | None            | Unsure                 |
| type             | Y           | N               | N               | DB migration           |
| subtype          | Y           | N               | N               | DB migration           |
| published        | Y           | N               | N               |                        |
| creationDate     | Y           | Y               | N               | Setter null            |
| modificationDate | Y           | Y               | N               | Setter null            |
| userOwner        | Y           | Y               | N               | Setter null            |
| userModification | Y           | Y               | N               | Setter null            |
| data             | Y           | Y               | N               | Setter null            |
| properties       | Y           | N               | N               | Setter and getter null |
| key              | Y           | Y               | Y               |                        |
| index            | Y           | Y               | Y               |                        |
```

**Comment by kingjia90** [comment]
Thank you for the feedback and this matrix looks great, could you provide a PR with all the solution applied?

**Comment by bramstroker** [comment]
@kingjia90 See #15943
